R/tools.R

```diff
@@ -182,3 +182,109 @@ check_sircovid_model <- function(sircovid_model) {
          call. = FALSE)
   }
 }
+
+
+##' Rotate strains, so that strain 1 becomes the sum of strains 1 and
+##' 2 and strain 2 is empty. Use this to allow sequential replacement
+##' of strains.
+##'
+##' @title Rotate strains
+##'
+##' @param state Model state
+##'
+##' @param info Model info
+##' @export
+rotate_strains <- function(state, info) {
+  if (is.null(dim(state))) {
+    stop("Expected a matrix or array for 'state'")
+  }
+  if (nrow(state) != info$len) {
+    stop(sprintf("Expected a matrix with %d rows for 'state'",
+                 info$len))
+  }
+
+  ## Push all state down into a common rank to avoid lots of boring
+  ## dimension arithmetic (e.g., we might have this structured by
+  ## region, or a single particle's state etc, but what matters is the
+  ## structure by particle only).
+  dim_orig <- dim(state)
+  if (length(dim(state)) > 2) {
+    state <- matrix(state, nrow(state))
+  }
+
+  ## Currently we only support one sort of move: Anyone who has been
+  ## infected in the past is moved to now be indexed by strain 1 (no
+  ## matter whether they had multiple infections before), and strain 2
+  ## will be completely empty.
+  strain_from_idx <- c(2, 3, 4)
+  strain_to_idx <- 1
+
+  ## TODO: someone please check this with the odin code. How would we
+  ## know if additional compartments are added that should be put
+  ## here?  In contrast with inflate_state which works by detecting
+  ## which states have changed size, we can't easily do that here.
+  ##
+  ## It's possible that something could be derived from the odin code
+  ## by looking for the expression that underpins each *variable* and
+  ## seeing if it contains n_strains. Rich can do this.
```

Comment:
I wondered about this and I don't know how to solve this, at the moment I did all this manually I'm afraid. Leaving this with you @richfitz I think...